NullPointerException occurs when I execute select statement via DBeaver

## Bug Report

### Which version of ShardingSphere did you use?

[3c11643](https://github.com/apache/shardingsphere/commit/3c11643483916e15dfd376c16b2ec3d4a0b2ae96)

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy

### Expected behavior

Execute select statement successfully via DBeaver

### Actual behavior

```
[INFO ] 2023-04-19 18:02:34.436 [Connection-4-ThreadExecutor] ShardingSphere-SQL - Logic SQL: /* ApplicationName=DBeaver 23.0.2 - SQLEditor <Console> */ select * from t_order
LIMIT 0, 200
[INFO ] 2023-04-19 18:02:34.436 [Connection-4-ThreadExecutor] ShardingSphere-SQL - Actual SQL: ds_0 ::: /* ApplicationName=DBeaver 23.0.2 - SQLEditor <Console> */ select * from t_order_0
LIMIT 0, 200
[INFO ] 2023-04-19 18:02:34.436 [Connection-4-ThreadExecutor] ShardingSphere-SQL - Actual SQL: ds_0 ::: /* ApplicationName=DBeaver 23.0.2 - SQLEditor <Console> */ select * from t_order_1
LIMIT 0, 200
[INFO ] 2023-04-19 18:02:34.436 [Connection-4-ThreadExecutor] ShardingSphere-SQL - Actual SQL: ds_1 ::: /* ApplicationName=DBeaver 23.0.2 - SQLEditor <Console> */ select * from t_order_0
LIMIT 0, 200
[INFO ] 2023-04-19 18:02:34.436 [Connection-4-ThreadExecutor] ShardingSphere-SQL - Actual SQL: ds_1 ::: /* ApplicationName=DBeaver 23.0.2 - SQLEditor <Console> */ select * from t_order_1
LIMIT 0, 200
[WARN ] 2023-04-19 18:03:08.012 [main-SendThread(localhost:2181)] org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 43104ms for session id 0x1000005660f01d0
[WARN ] 2023-04-19 18:03:08.014 [main-SendThread(localhost:2181)] org.apache.zookeeper.ClientCnxn - Session 0x1000005660f01d0 for server localhost/127.0.0.1:2181, Closing socket connection. Attempting reconnect except it is a SessionExpiredException.
org.apache.zookeeper.ClientCnxn$SessionTimeoutException: Client session timed out, have not heard from server in 43104ms for session id 0x1000005660f01d0
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1257)
[ERROR] 2023-04-19 18:03:43.869 [Connection-3-ThreadExecutor] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
java.lang.NullPointerException: null
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.FederationTranslatableTable.getColumnType(FederationTranslatableTable.java:120)
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.TranslatableTableScan.traverseRexCall(TranslatableTableScan.java:205)
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.TranslatableTableScan.traverseRexCall(TranslatableTableScan.java:209)
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.TranslatableTableScan.traverseRexCall(TranslatableTableScan.java:209)
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.TranslatableTableScan.generateColumnMap(TranslatableTableScan.java:195)
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.TranslatableTableScan.addFilter(TranslatableTableScan.java:187)
	at org.apache.shardingsphere.sqlfederation.optimizer.metadata.translatable.TranslatableTableScan.explainTerms(TranslatableTableScan.java:127)
	at org.apache.calcite.rel.AbstractRelNode.getDigestItems(AbstractRelNode.java:415)
	at org.apache.calcite.rel.AbstractRelNode.deepHashCode(AbstractRelNode.java:397)
	at org.apache.calcite.rel.AbstractRelNode$InnerRelDigest.hashCode(AbstractRelNode.java:449)
	at java.base/java.util.HashMap.hash(HashMap.java:339)
	at java.base/java.util.HashMap.get(HashMap.java:552)
	at org.apache.calcite.plan.hep.HepPlanner.addRelToGraph(HepPlanner.java:829)
	at org.apache.calcite.plan.hep.HepPlanner.applyTransformationResults(HepPlanner.java:751)
	at org.apache.calcite.plan.hep.HepPlanner.applyRule(HepPlanner.java:559)
	at org.apache.calcite.plan.hep.HepPlanner.applyRules(HepPlanner.java:420)
	at org.apache.calcite.plan.hep.HepPlanner.executeEndGroup(HepPlanner.java:360)
	at org.apache.calcite.plan.hep.HepInstruction$EndGroup$State.execute(HepInstruction.java:349)
	at org.apache.calcite.plan.hep.HepPlanner.lambda$executeProgram$0(HepPlanner.java:211)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:405)
	at org.apache.calcite.plan.hep.HepPlanner.executeProgram(HepPlanner.java:210)
	at org.apache.calcite.plan.hep.HepProgram$State.execute(HepProgram.java:118)
	at org.apache.calcite.plan.hep.HepPlanner.executeProgram(HepPlanner.java:205)
	at org.apache.calcite.plan.hep.HepPlanner.findBestExp(HepPlanner.java:191)
	at org.apache.shardingsphere.sqlfederation.optimizer.SQLOptimizeEngine.optimizeWithRBO(SQLOptimizeEngine.java:59)
	at org.apache.shardingsphere.sqlfederation.optimizer.SQLOptimizeEngine.optimize(SQLOptimizeEngine.java:52)
	at org.apache.shardingsphere.sqlfederation.advanced.AdvancedSQLFederationExecutor.execute(AdvancedSQLFederationExecutor.java:144)
	at org.apache.shardingsphere.sqlfederation.advanced.AdvancedSQLFederationExecutor.executeQuery(AdvancedSQLFederationExecutor.java:114)
	at org.apache.shardingsphere.proxy.backend.connector.DatabaseConnector.doExecuteFederation(DatabaseConnector.java:298)
	at org.apache.shardingsphere.proxy.backend.connector.DatabaseConnector.execute(DatabaseConnector.java:205)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.execute(MySQLComQueryPacketExecutor.java:92)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.doExecuteCommand(CommandExecutorTask.java:116)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:111)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
[INFO ] 2023-04-19 18:03:43.878 [main-EventThread] o.a.c.f.state.ConnectionStateManager - State change: SUSPENDED
[INFO ] 2023-04-19 18:03:45.314 [main-SendThread(localhost:2181)] org.apache.zookeeper.ClientCnxn - Opening socket connection to server localhost/127.0.0.1:2181.
```

### Reason analyze (If you can)

DBeaver will execute `SELECT * FROM INFORMATION_SCHEMA.COLUMNS`:

<img width="1269" alt="image" src="https://user-images.githubusercontent.com/10829171/233044025-7eebbb1e-3781-40d0-8a91-afd919369d7e.png">

And sql federation doesn't get ShardingSphereColumn because of upper case.

<img width="1183" alt="image" src="https://user-images.githubusercontent.com/10829171/233044301-9d781598-08e6-4b7f-92f8-f6d5049c1ac7.png">

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/merge/dal/show/ShowCreateTableMergedResult.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/merge/dal/show/ShowCreateTableMergedResult.java
@@ -51,7 +51,7 @@ public final class ShowCreateTableMergedResult extends LogicTablesMergedResult {
             String actualIndexName = IndexMetaDataUtils.getActualIndexName(each, actualTableName);
             memoryResultSetRow.setCell(2, memoryResultSetRow.getCell(2).toString().replace(actualIndexName, each));
         }
-        for (Entry<String, ShardingSphereConstraint> entry : table.getConstrains().entrySet()) {
+        for (Entry<String, ShardingSphereConstraint> entry : table.getConstraints().entrySet()) {
             String actualIndexName = IndexMetaDataUtils.getActualIndexName(entry.getKey(), actualTableName);
             memoryResultSetRow.setCell(2, memoryResultSetRow.getCell(2).toString().replace(actualIndexName, entry.getKey()));
             Optional<TableRule> tableRule = shardingRule.findTableRule(entry.getValue().getReferencedTableName());

--- a/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/model/ShardingSphereTable.java
+++ b/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/model/ShardingSphereTable.java
@@ -44,7 +44,7 @@ public final class ShardingSphereTable {
     
     private final Map<String, ShardingSphereIndex> indexes;
     
-    private final Map<String, ShardingSphereConstraint> constrains;
+    private final Map<String, ShardingSphereConstraint> constraints;
     
     private final List<String> columnNames = new ArrayList<>();
     
@@ -56,17 +56,17 @@ public final class ShardingSphereTable {
         this("", Collections.emptyList(), Collections.emptyList(), Collections.emptyList());
     }
     
-    public ShardingSphereTable(final String name, final Collection<ShardingSphereColumn> columnList,
-                               final Collection<ShardingSphereIndex> indexList, final Collection<ShardingSphereConstraint> constraintList) {
+    public ShardingSphereTable(final String name, final Collection<ShardingSphereColumn> columns,
+                               final Collection<ShardingSphereIndex> indexes, final Collection<ShardingSphereConstraint> constraints) {
         this.name = name;
-        columns = getColumns(columnList);
-        indexes = getIndexes(indexList);
-        constrains = getConstrains(constraintList);
+        this.columns = getColumns(columns);
+        this.indexes = getIndexes(indexes);
+        this.constraints = getConstraints(constraints);
     }
     
-    private Map<String, ShardingSphereColumn> getColumns(final Collection<ShardingSphereColumn> columnList) {
-        Map<String, ShardingSphereColumn> result = new LinkedHashMap<>(columnList.size(), 1);
-        for (ShardingSphereColumn each : columnList) {
+    private Map<String, ShardingSphereColumn> getColumns(final Collection<ShardingSphereColumn> columns) {
+        Map<String, ShardingSphereColumn> result = new LinkedHashMap<>(columns.size(), 1);
+        for (ShardingSphereColumn each : columns) {
             String lowerColumnName = each.getName().toLowerCase();
             result.put(lowerColumnName, each);
             columnNames.add(each.getName());
@@ -80,19 +80,29 @@ public final class ShardingSphereTable {
         return result;
     }
     
-    private Map<String, ShardingSphereIndex> getIndexes(final Collection<ShardingSphereIndex> indexList) {
-        Map<String, ShardingSphereIndex> result = new LinkedHashMap<>(indexList.size(), 1);
-        for (ShardingSphereIndex each : indexList) {
+    private Map<String, ShardingSphereIndex> getIndexes(final Collection<ShardingSphereIndex> indexes) {
+        Map<String, ShardingSphereIndex> result = new LinkedHashMap<>(indexes.size(), 1);
+        for (ShardingSphereIndex each : indexes) {
             result.put(each.getName().toLowerCase(), each);
         }
         return result;
     }
     
-    private Map<String, ShardingSphereConstraint> getConstrains(final Collection<ShardingSphereConstraint> constraintList) {
-        Map<String, ShardingSphereConstraint> result = new LinkedHashMap<>(constraintList.size(), 1);
-        for (ShardingSphereConstraint each : constraintList) {
+    private Map<String, ShardingSphereConstraint> getConstraints(final Collection<ShardingSphereConstraint> constraints) {
+        Map<String, ShardingSphereConstraint> result = new LinkedHashMap<>(constraints.size(), 1);
+        for (ShardingSphereConstraint each : constraints) {
             result.put(each.getName().toLowerCase(), each);
         }
         return result;
+    }
+    
+    /**
+     * Get table meta data via column name.
+     *
+     * @param columnName column name
+     * @return table meta data
+     */
+    public ShardingSphereColumn getColumn(final String columnName) {
+        return columns.get(columnName.toLowerCase());
     }
 }

--- a/infra/common/src/main/java/org/apache/shardingsphere/infra/yaml/schema/swapper/YamlTableSwapper.java
+++ b/infra/common/src/main/java/org/apache/shardingsphere/infra/yaml/schema/swapper/YamlTableSwapper.java
@@ -43,7 +43,7 @@ public final class YamlTableSwapper implements YamlConfigurationSwapper<YamlShar
         YamlShardingSphereTable result = new YamlShardingSphereTable();
         result.setColumns(swapYamlColumns(table.getColumns()));
         result.setIndexes(swapYamlIndexes(table.getIndexes()));
-        result.setConstraints(swapYamlConstraints(table.getConstrains()));
+        result.setConstraints(swapYamlConstraints(table.getConstraints()));
         result.setName(table.getName());
         return result;
     }

--- a/infra/context/src/main/java/org/apache/shardingsphere/infra/context/refresher/type/index/AlterIndexStatementSchemaRefresher.java
+++ b/infra/context/src/main/java/org/apache/shardingsphere/infra/context/refresher/type/index/AlterIndexStatementSchemaRefresher.java
@@ -68,7 +68,7 @@ public final class AlterIndexStatementSchemaRefresher implements MetaDataRefresh
     
     private ShardingSphereTable newShardingSphereTable(final ShardingSphereTable table) {
         ShardingSphereTable result = new ShardingSphereTable(table.getName(), new LinkedHashMap<>(table.getColumns()),
-                new LinkedHashMap<>(table.getIndexes()), new LinkedHashMap<>(table.getConstrains()));
+                new LinkedHashMap<>(table.getIndexes()), new LinkedHashMap<>(table.getConstraints()));
         result.getColumnNames().addAll(table.getColumnNames());
         result.getVisibleColumns().addAll(table.getVisibleColumns());
         result.getPrimaryKeyColumns().addAll(table.getPrimaryKeyColumns());

--- a/infra/context/src/main/java/org/apache/shardingsphere/infra/context/refresher/type/index/CreateIndexStatementSchemaRefresher.java
+++ b/infra/context/src/main/java/org/apache/shardingsphere/infra/context/refresher/type/index/CreateIndexStatementSchemaRefresher.java
@@ -54,7 +54,7 @@ public final class CreateIndexStatementSchemaRefresher implements MetaDataRefres
     
     private ShardingSphereTable newShardingSphereTable(final ShardingSphereTable table) {
         ShardingSphereTable result = new ShardingSphereTable(table.getName(), new LinkedHashMap<>(table.getColumns()),
-                new LinkedHashMap<>(table.getIndexes()), new LinkedHashMap<>(table.getConstrains()));
+                new LinkedHashMap<>(table.getIndexes()), new LinkedHashMap<>(table.getConstraints()));
         result.getColumnNames().addAll(table.getColumnNames());
         result.getVisibleColumns().addAll(table.getVisibleColumns());
         result.getPrimaryKeyColumns().addAll(table.getPrimaryKeyColumns());

--- a/infra/context/src/main/java/org/apache/shardingsphere/infra/context/refresher/type/index/DropIndexStatementSchemaRefresher.java
+++ b/infra/context/src/main/java/org/apache/shardingsphere/infra/context/refresher/type/index/DropIndexStatementSchemaRefresher.java
@@ -68,7 +68,7 @@ public final class DropIndexStatementSchemaRefresher implements MetaDataRefreshe
     
     private ShardingSphereTable newShardingSphereTable(final ShardingSphereTable table) {
         ShardingSphereTable result = new ShardingSphereTable(table.getName(), new LinkedHashMap<>(table.getColumns()),
-                new LinkedHashMap<>(table.getIndexes()), new LinkedHashMap<>(table.getConstrains()));
+                new LinkedHashMap<>(table.getIndexes()), new LinkedHashMap<>(table.getConstraints()));
         result.getColumnNames().addAll(table.getColumnNames());
         result.getVisibleColumns().addAll(table.getVisibleColumns());
         result.getPrimaryKeyColumns().addAll(table.getPrimaryKeyColumns());

--- a/kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/FederationTranslatableTable.java
+++ b/kernel/sql-federation/optimizer/src/main/java/org/apache/shardingsphere/sqlfederation/optimizer/metadata/translatable/FederationTranslatableTable.java
@@ -35,7 +35,6 @@ import org.apache.calcite.schema.Statistic;
 import org.apache.calcite.schema.TranslatableTable;
 import org.apache.calcite.schema.impl.AbstractTable;
 import org.apache.shardingsphere.infra.database.type.DatabaseType;
-import org.apache.shardingsphere.infra.metadata.database.schema.model.ShardingSphereColumn;
 import org.apache.shardingsphere.infra.metadata.database.schema.model.ShardingSphereTable;
 import org.apache.shardingsphere.infra.util.exception.external.sql.type.generic.UnsupportedSQLOperationException;
 import org.apache.shardingsphere.sqlfederation.optimizer.executor.TableScanExecutor;
@@ -44,8 +43,6 @@ import org.apache.shardingsphere.sqlfederation.optimizer.metadata.statistic.Fede
 import org.apache.shardingsphere.sqlfederation.optimizer.util.SQLFederationDataTypeUtils;
 
 import java.lang.reflect.Type;
-import java.util.List;
-import java.util.Map;
 
 /**
  * Federation translatable table.
@@ -114,10 +111,8 @@ public final class FederationTranslatableTable extends AbstractTable implements 
      * @return column data type
      */
     public int getColumnType(final int index) {
-        List columnNames = table.getVisibleColumns();
-        Map<String, ShardingSphereColumn> columnMap = table.getColumns();
-        ShardingSphereColumn column = columnMap.get(columnNames.get(index));
-        return column.getDataType();
+        String columnName = table.getColumnNames().get(index);
+        return table.getColumn(columnName).getDataType();
     }
     
     @Override

